File: WebClient SDK/src/default/ui/tagDemoCardActionExtension.ts
import userSession from 'common/utility/userSession';
import { runInAction } from 'mobx';
import moment from 'moment';
import { CardStoreMode } from 'tessa/cards/cardStoreMode';
import { Tag } from 'tessa/tags/tag';
import { TagInfo } from 'tessa/tags/tagInfo';
import { CardUIExtension } from 'tessa/ui/cards/cardUIExtension';
import { ICardUIExtensionContext } from 'tessa/ui/cards/cardUIExtensionContext';
import { ICardEditorModel } from 'tessa/ui/cards/interfaces';
import { CardTagViewModel } from 'tessa/ui/tags/cardTagViewModel';
import { selectOrCreateTag } from 'tessa/ui/tags/tagHelper';
import { TagInfoModel } from 'tessa/ui/tags/tagInfoModel';
import { TagManager } from 'tessa/ui/tags/tagManager';
import { showConfirm } from 'tessa/ui/tessaDialog/show';
import { showNotEmpty } from 'tessa/ui/tessaDialog/showNotEmpty';
import { UIContext } from 'tessa/ui/uiContext';

export class TagDemoCardActionExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext): Promise<void> {
    const cardEditor = context.uiContext.cardEditor!;
    cardEditor.toolbar.canHaveTags = true;
    const tags = await TagManager.instance.getTags(context.card.id);
    cardEditor.toolbar.canHaveTags = true;
    runInAction(() => {
      cardEditor.toolbar.tags.length = 0;
      for (const tagInfo of tags) {
        this.addTagToToolbar(context.card.id, cardEditor, tagInfo);
      }
    });
    cardEditor.toolbar.clickAddTag = async () => {
      const cardEditor = UIContext.current.cardEditor;
      if (!cardEditor?.cardModel) {
        return;
      }

      const cardId = cardEditor.cardModel.card.id;

      if (
        cardEditor.cardModel.card.storeMode === CardStoreMode.Insert ||
        (await cardEditor.cardModel.hasChanges())
      ) {
        if (await showConfirm('$Tags_SaveCardToAddTag')) {
          const saveResult = await cardEditor.cardModel.saveAsync();
          if (!saveResult.isSuccessful) {
            return;
          }
        } else {
          return;
        }
      }

      const tagInfo = await selectOrCreateTag(cardId);

      if (tagInfo && !cardEditor.toolbar.tags.some(x => x.id === tagInfo.id && !x.isDeleted)) {
        const tag = new Tag();
        tag.tagId = tagInfo.id;
        tag.cardId = context.card.id;
        tag.userId = userSession.UserID;
        tag.setAt = moment.utc().format();
        const result = await TagManager.instance.storeTag(tag);
        const tagStored = result.isSuccessful;
        await showNotEmpty(result);
        if (tagStored) {
          this.addTagToToolbar(cardId, cardEditor, tagInfo);
        }
      }
    };
  }

  private addTagToToolbar(cardId: guid, cardEditor: ICardEditorModel, tag: TagInfo) {
    // Проверяем, не был ли тег удален перед этим.
    // Если был, то восстанавливаем его, а не добавляем новый.
    const deletedTag = cardEditor.toolbar.tags.find(x => x.id == tag.id && x.isDeleted);
    if (deletedTag) {
      runInAction(() => {
        deletedTag.isDeleted = false;
      });
    } else {
      const tagInfoModel = new TagInfoModel(tag, cardId);
      const viewTagInfoModel = new CardTagViewModel(tagInfoModel);
      runInAction(() => {
        cardEditor.toolbar.tags.push(viewTagInfoModel);
      });
    }
  }

  public shouldExecute(_context: ICardUIExtensionContext): boolean {
    return _context.card.typeId === 'd0006e40-a342-4797-8d77-6501c4b7c4ac'; // Карточка автомобиля.
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/TagDemoCardActionExtension.cs
#nullable enable

using System;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Tags;
using Tessa.Platform.Runtime;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tags;
using Tessa.UI.Tags.ViewModels;
using Tessa.Localization;
using Tessa.Cards;

namespace Tessa.Extensions.Default.Client.UI
{
    /// <summary>
    /// Пример UI-расширения для работы тегов в карточке.
    /// </summary>
    public sealed class TagDemoCardActionExtension :
        CardUIExtension
    {
        #region Constructors

        public TagDemoCardActionExtension(
            IUIHost uiHost,
            IAdvancedCardDialogManager cardDialogManager,
            CreateDialogFormFuncAsync createDialogFormFuncAsync,
            ITagManager tagManager,
            ITagInfoModelFactory tagInfoModelFactory,
            ISession session)
        {
            this.uiHost = NotNullOrThrow(uiHost);
            this.cardDialogManager = NotNullOrThrow(cardDialogManager);
            this.createDialogFormFuncAsync = NotNullOrThrow(createDialogFormFuncAsync);
            this.tagManager = NotNullOrThrow(tagManager);
            this.tagInfoModelFactory = NotNullOrThrow(tagInfoModelFactory);
            this.session = NotNullOrThrow(session);
        }

        #endregion

        #region Fields

        private readonly IUIHost uiHost;
        private readonly IAdvancedCardDialogManager cardDialogManager;
        private readonly CreateDialogFormFuncAsync createDialogFormFuncAsync;
        private readonly ITagManager tagManager;
        private readonly ITagInfoModelFactory tagInfoModelFactory;
        private readonly ISession session;

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override async Task Initialized(ICardUIExtensionContext context)
        {
            if (context.Card is Card card &&
                context.UIContext.CardEditor is ICardEditorModel cardEditor)
            {
                cardEditor.Toolbar.CanHaveTags = true;
                cardEditor.Toolbar.TagsModel.Tags.Clear();
                var allTagsInfo = await this.tagManager.GetTagsAsync(context.Card.ID, context.ValidationResult, context.CancellationToken);

                if (!context.ValidationResult.IsSuccessful())
                {
                    return;
                }

                if (allTagsInfo is null)
                {
                    cardEditor.Toolbar.CanHaveTags = false;
                    return;
                }

                cardEditor.Toolbar.CanHaveTags = true;
                cardEditor.Toolbar.TagsModel.Tags.Clear();
                foreach (var tagInfo in allTagsInfo)
                {
                    this.AddTagToToolbar(context.Card, cardEditor, tagInfo);
                }

                if (cardEditor.Toolbar.RightTagControl is AddTagButtonViewModel addTagButtonViewModel)
                {
                    addTagButtonViewModel.AddTagCommand = new DelegateCommand(async _ =>
                    {

                        if (cardEditor is null
                            || cardEditor.CardModel is null)
                        {
                            return;
                        }
                        var cardID = cardEditor.CardModel.Card.ID;

                        // Если в карточке были изменения, необходимо сохранить ее перед добавлением тега. 

                        if (cardEditor.CardModel.Card.StoreMode == CardStoreMode.Insert || 
                            await cardEditor.CardModel.HasChangesAsync())
                        {
                            if (await TessaDialog.ConfirmAsync(await LocalizationManager.GetStringAsync("Tags_SaveCardToAddTag")))
                            {
                                var saveResult = await cardEditor.CardModel.SaveAsync();
                                if (!saveResult.Success)
                                {
                                    return;
                                }
                            }
                            else
                            {
                                return;
                            }
                        }

                        var tagInfo = await TagUIExtensions.SelectOrCreateTagAsync(
                            this.createDialogFormFuncAsync,
                            this.uiHost,
                            this.cardDialogManager,
                            cardID);

                        if (!(tagInfo is not null &&
                            !cardEditor.Toolbar.TagsModel.Tags.Any(x => x.Id == tagInfo.ID && !x.IsDeleted)))
                        {
                            return;
                        }

                        var tag = new Tag()
                        {
                            SetAt = DateTime.UtcNow,
                            CardID = context.Card.ID,
                            TagID = tagInfo.ID,
                            UserID = this.session.User.ID
                        };

                        await this.tagManager.AddTagAsync(tag, null, context.ValidationResult);
                        if (context.ValidationResult.IsSuccessful())
                        {
                            this.AddTagToToolbar(cardEditor.CardModel.Card, cardEditor, tagInfo);
                        }
                    });
                }
            }
        }

        #endregion

        #region Private Methods

        private void AddTagToToolbar(Card card, ICardEditorModel cardEditor, TagInfo tag)
        {
            // Проверяем, не был ли тег удален перед этим.
            // Если был, то восстанавливаем его, а не добавляем новый.
            var deletedTag = cardEditor.Toolbar.TagsModel.Tags.FirstOrDefault(x => x.Id == tag.ID);
            if (deletedTag is not null)
            {
                deletedTag.IsDeleted = false;
            }
            else
            {
                var tagInfoModel = this.tagInfoModelFactory.CreateModel(tag, card);
                var viewTagInfoModel = new CardTagViewModel(tagInfoModel);
                cardEditor.Toolbar.TagsModel.Tags.Add(viewTagInfoModel);
            }
        }

        #endregion
    }
}
